control/server.py

```diff
@@ -773,18 +773,56 @@ def _stop_spdk(self):
                 self.logger.exception(f"An error occurred while removing RPC "
                                       f"socket {self.spdk_rpc_socket_path}")
 
+    def _terminate_discovery(self, pid):
+        def is_running(pid):
+            try:
+                os.kill(pid, 0)
+                return True
+            except ProcessLookupError:
+                return False
+            except PermissionError:
+                self.logger.exception(f"Permission denied when checking status of discovery {pid}")
+                return True
+
+        def wait_for_exit(pid):
+            WAIT_INTERVAL_SEC = 0.1
+            MAX_WAIT_ATTEMPTS = 10
+            for _ in range(MAX_WAIT_ATTEMPTS):
+                if not is_running(pid):
+                    return True
+                time.sleep(WAIT_INTERVAL_SEC)
+            return False
+
+        try:
+            # discovery service selector loop should exit
+            # due to KeyboardInterrupt exception on SIGINT signal
+            signals = [signal.SIGINT, signal.SIGTERM, signal.SIGKILL]
+            for sig in signals:
+                self.logger.info(f"Sending signal {sig.name} to Discovery service process {pid}")
+                os.kill(pid, sig)
+                if wait_for_exit(pid):
+                    break
+            else:
+                self.logger.warning(
+                    f"Discovery service process {pid} did not exit after all signals."
+                )
+
+            return not is_running(pid)
```

Comment:
```suggestion
            signals = [signal.SIGINT, signal.SIGTERM, signal.SIGKILL]
            for sig in signals:
                self.logger.info(f"Sending signal {sig.name} to Discovery service process {pid}")
                os.kill(pid, sig)
                if wait_for_exit(pid):
                    return True
            self.logger.warning(
                f"Discovery service process {pid} did not exit after all signals."
            )
            return False
```
I wonder if my suggestion is easier to comprehend since I think `for else` syntax is a bit confusing :)  
just double check I didn't create any bugs haha